src/reference_implementations/prompting_vector_llms/llm_prompting_examples/llm_prompt_ag_news.ipynb

```diff
@@ -0,0 +1,1107 @@
+      "Description: Reuters - British scientists said on Wednesday they had received permission to clone human embryos for medical research, in what they believe to be the first such license to be granted in Europe.\n",
+      "Category (World, Sports, Business, Technology): Technology\n",
+      "\n",
+      "Title: Telecom lifts first quarter net profit 19pc\n",
+      "Description: Telecom Corp today reported its September first quarter net profit rose 19 per cent to $193 million. The profit bettered analysts #39; average forecasts of $185m.\n",
+      "Category (World, Sports, Business, Technology):\n"
+     ]
+    }
+   ],
+   "source": [
+    "prompt_template_postfix = \"Category (World, Sports, Business, Technology):\"\n",
+    "sample_texts = [\n",
+    "    f\"{prompt_demonstrations}{model_input_text}\\n{prompt_template_postfix}\"\n",
+    "    for model_input_text in model_input_texts[0:3]\n",
+    "]\n",
+    "print(f\"Prompt Example\\n{sample_texts[0]}\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Technology\n",
+      "==================================\n",
+      "Technology\n",
+      "==================================\n",
+      "Sports\n",
+      "==================================\n"
+     ]
+    }
+   ],
+   "source": [
+    "generation = model.generate(sample_texts, short_generation_config)\n",
+    "for text in generation.generation[\"sequences\"]:\n",
+    "    print(text)\n",
+    "    print(\"==================================\")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Few-shot learning definitely helps a lot! We'll measure accuracy on a sample of the AG news dataset below. However, there is nothing stoping the model from not selecting our labels. So can we do better? We can work around this by understanding the likelihood of our labels from the models perspective. This will also allow us to use zero-shot learning, even when the model doesn't seem to want to respond in the way we expect."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "'output'"
+      ]
+     },
+     "execution_count": 16,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# We're interested in the activations from the last layer of the model, because this will allow us to calculate the\n",
+    "# likelihoods\n",
+    "last_layer_name = model.module_names[-1]\n",
+    "last_layer_name"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The last layer of the model is analogous to the probabilities of each token in the model vocabulary. That is, it is the conditional probability\n",
```

Comment:
I think we can rewrite this paragraph a bit better. For example: 

"last layer of the model" -> "last layer activations of the model"

"we get back a 32K vector..." -> "we get back a vector of dimension 32000 (the vocabulary size of Llama-2) which represents the probability distribution of  $y_{t+1}$ over the vocabulary..."